clippy_lints/src/inherent_impl.rs

```diff
@@ -56,33 +57,41 @@ impl<'tcx> LateLintPass<'tcx> for MultipleInherentImpl {
             ..
         }) = item.kind
         {
-            // Remember for each inherent implementation encountered its span and generics
+            // Remember for each inherent implementation encountered its id and span,
             // but filter out implementations that have generic params (type or lifetime)
             // or are derived from a macro
             if !in_macro(item.span) && generics.params.is_empty() {
-                self.impls.insert(item.def_id.to_def_id(), item.span);
+                self.impls.push((item.def_id, item.span));
             }
         }
     }
 
     fn check_crate_post(&mut self, cx: &LateContext<'tcx>, krate: &'tcx Crate<'_>) {
         if !krate.items.is_empty() {
-            // Retrieve all inherent implementations from the crate, grouped by type
-            for impls in cx.tcx.crate_inherent_impls(def_id::LOCAL_CRATE).inherent_impls.values() {
-                // Filter out implementations that have generic params (type or lifetime)
-                let mut impl_spans = impls.iter().filter_map(|impl_def| self.impls.get(impl_def));
-                if let Some(initial_span) = impl_spans.next() {
-                    impl_spans.for_each(|additional_span| {
+            // Split the collected impls by type
+            let mut type_map = FxIndexMap::default();
+            for (ty, span) in self.impls.iter().map(|&(did, span)| (cx.tcx.type_of(did), span)) {
+                match type_map.entry(ty) {
+                    Entry::Vacant(e) => {
+                        e.insert(vec![span]);
+                    },
+                    Entry::Occupied(mut e) => e.get_mut().push(span),
```

Comment:
Thought of one more suggestion if you like. You could lint right here instead of adding to the map. Then it could be a HashMap with only the first impl per type.

Reply:
That would be better. The only difference would be the order things are printed. It's currently sorted by type, then declaration order. This would make it solely by declaration order.

Reply:
That okay.